pkg/cache/v3/snapshot.go

```diff
@@ -94,64 +35,30 @@ type Snapshot struct {
 }
 
 // NewSnapshot creates a snapshot from response types and a version.
-func NewSnapshot(version string,
-	endpoints []types.Resource,
-	clusters []types.Resource,
-	routes []types.Resource,
-	listeners []types.Resource,
-	runtimes []types.Resource,
-	secrets []types.Resource,
-	extensionConfigs []types.Resource) Snapshot {
-	return NewSnapshotWithResources(version, SnapshotResources{
-		Endpoints:        endpoints,
-		Clusters:         clusters,
-		Routes:           routes,
-		Listeners:        listeners,
-		Runtimes:         runtimes,
-		Secrets:          secrets,
-		ExtensionConfigs: extensionConfigs,
-	})
-}
-
-// SnapshotResources contains the resources to construct a snapshot from.
-type SnapshotResources struct {
-	Endpoints        []types.Resource
-	Clusters         []types.Resource
-	Routes           []types.Resource
-	Listeners        []types.Resource
-	Runtimes         []types.Resource
-	Secrets          []types.Resource
-	ExtensionConfigs []types.Resource
-}
-
-// NewSnapshotWithResources creates a snapshot from response types and a version.
-func NewSnapshotWithResources(version string, resources SnapshotResources) Snapshot {
+func NewSnapshot(version string, resources map[string][]types.Resource) Snapshot {
```

Comment:
So make it clear that this map is keyed on resource type, consider switching to a typedef. This would make it easier to keep straight which keys are resource type names, and which are resource object names.

```Go
type ResourceType string

const EndpointType = ResourceType(apiTypePrefix + "envoy.config.endpoint.v3.ClusterLoadAssignment")
```

Reply:
Hm yea I like that. My only thought to not do something like that is to keep the generic ability of any type, not just what is defined in the resources file. But I think yea it's probably worth doing some sort of typedef and then users can just create their own custom types

Reply:
I'll give it a shot and see how it goes

Reply:
I think at the very least it's worth adding a doc string noting the key is the resource type

Reply:
Yeh if you make it a type, clients can still provide their own values of that type.

Reply:
So this is kinda awkward since the type url in a `DiscoveryRequest` is of type `string`. If I were to require a type alias here we'd have to do a lot of type casting back and forth between string

Reply:
Maybe you can cast it once at the API boundary (and verify the type in the protobuf registry)?

Reply:
That seemed to work out nicely @jpeach thanks :)